### PR TITLE
Replace hashtables with generic dictionaries in ToolTip

### PR DESCRIPTION
**Pull request by gpetrou**
## Proposed changes

- Replace hashtables with generic dictionaries in ToolTip.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/7294)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolTip.cs [hunk 1]
```diff
@@ -502,8 +501,7 @@ private Control TopLevelControl
                 }
 
                 Control currentTopLevel = null;
-                var regions = new Control[_tools.Keys.Count];
-                _tools.Keys.CopyTo(regions, 0);
+                var regions = _tools.Keys.ToArray();
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolTip.cs:504]: Please update using the explicit type instead `var` in such cases. Ditto below.

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolTip.cs [hunk 2]
```diff
@@ -1167,8 +1161,7 @@ private void RecreateHandle()
         /// </summary>
         public void RemoveAll()
         {
-            Control[] regions = new Control[_tools.Keys.Count];
-            _tools.Keys.CopyTo(regions, 0);
+            Control[] regions = _tools.Keys.ToArray();
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolTip.cs:1164]:
```suggestion
            Control[] controls = _tools.Keys.ToArray();
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolTip.cs (hunk 1 above)
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolTip.cs:504]:
```suggestion
                Control[] controls = _tools.Keys.ToArray();
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolTip.cs [hunk 3]
```diff
@@ -1811,8 +1816,7 @@ private void BaseFormDeactivate(object sender, EventArgs e)
 
         private void HideAllToolTips()
         {
-            Control[] ctls = new Control[_owners.Values.Count];
-            _owners.Values.CopyTo(ctls, 0);
+            Control[] ctls = _owners.Values.ToArray();
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolTip.cs:1819]:
```suggestion
            Control[] controls = _owners.Values.ToArray();
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolTip.cs [hunk 4]
```diff
@@ -1540,10 +1532,23 @@ internal void ShowKeyboardToolTip(string text, IKeyboardToolTip tool, int durati
...
+                TipInfo tipInfo;
+                if (tool is Control toolAsControl)
                 {
-                    tipInfo.Position = new Point(pointX, pointY);
+                    if (!_tools.TryGetValue(toolAsControl, out tipInfo))
+                    {
+                        if (tool.GetOwnerWindow() is Control ownerWindowAsControl)
+                        {
+                            if (_tools.TryGetValue(ownerWindowAsControl, out tipInfo))
+                            {
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolTip.cs:1543]: Merge together

**Review by RussKie — COMMENTED** [review]
👍

**Review by Tanya-Solyanik — APPROVED** [review]
Looks good!

**Pull request merged**